src/Swarm/Game/Failure.hs

```diff
@@ -38,26 +47,54 @@ data LoadingFailure
   = DoesNotExist Entry
   | EntryNot Entry
   | CanNotParse ParseException
+  | Duplicate AssetData Text
   | CustomMessage Text
   deriving (Show)
 
-tShowLow :: Show a => a -> Text
-tShowLow = T.toLower . T.pack . show
+data OrderFileWarning
+  = NoOrderFile
+  | MissingFiles [FilePath]
+  | DanglingFiles [FilePath]
+  deriving (Eq, Show)
+
+------------------------------------------------------------
+-- Pretty-printing
+
+prettyAssetData :: AssetData -> Text
+prettyAssetData NameGeneration = "name generation data"
+prettyAssetData AppAsset = "data assets"
+prettyAssetData d = showLowT d
 
-tShow :: Show a => a -> Text
-tShow = T.pack . show
+prettyAsset :: Asset -> Text
+prettyAsset (Data ad) = prettyAssetData ad
+prettyAsset a = showLowT a
 
 prettyLoadingFailure :: LoadingFailure -> Text
 prettyLoadingFailure = \case
-  DoesNotExist e -> "The " <> tShowLow e <> " is missing!"
-  EntryNot e -> "The entry is not a " <> tShowLow e <> "!"
-  CanNotParse p -> "Parse failure:\n" <> T.pack (indent 8 $ prettyPrintParseException p)
+  DoesNotExist e -> "The " <> showLowT e <> " is missing!"
+  EntryNot e -> "The entry is not a " <> showLowT e <> "!"
+  CanNotParse p -> "Parse failure:\n" <> into @Text (indent 8 $ prettyPrintParseException p)
+  Duplicate thing duped -> "Duplicate " <> showLowT thing <> ": " <> quote duped
   CustomMessage m -> m
  where
   indent n = unlines . map (replicate n ' ' ++) . lines
 
--- | Pretty print system failure.
 prettyFailure :: SystemFailure -> Text
 prettyFailure = \case
   AssetNotLoaded a fp l ->
-    T.unwords ["Failed to acquire", tShowLow a, tShow fp, "from path", quote $ T.pack fp] <> ": " <> prettyLoadingFailure l
+    T.unwords ["Failed to acquire", prettyAsset a, "from path", quote $ into @Text fp] <> ": " <> prettyLoadingFailure l
+  ScenarioNotFound s ->
+    "Scenario not found: " <> into @Text s
+  OrderFileWarning orderFile w ->
+    "Warning: while processing " <> into @Text orderFile <> ":\n" <> T.unlines (prettyOrderFileWarning w)
+  CustomFailure m -> m
+
+prettyOrderFileWarning :: OrderFileWarning -> [Text]
+prettyOrderFileWarning = \case
+  NoOrderFile -> ["  File not found; using alphabetical order"]
+  MissingFiles missing ->
+    "  Files not listed will be ignored:"
+      : map (("  - " <>) . into @Text) missing
```

Comment:
Opportunity to use [`BulletList`](https://github.com/swarm-game/swarm/blob/c69d76edf3c4547c363bdb2bc387a7fa773e3b12/src/Swarm/Language/Pretty.hs#L75-L78)?